validator_client/src/http_api/mod.rs

```diff
@@ -1226,38 +1247,38 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
 /// Executes `func` in blocking tokio task (i.e., where long-running tasks are permitted).
 /// JSON-encodes the return value of `func`, using the `signer` function to produce a signature of
 /// those bytes.
-pub async fn blocking_signed_json_task<S, F, T>(
-    signer: S,
-    func: F,
-) -> Result<impl warp::Reply, warp::Rejection>
+pub async fn blocking_signed_json_task<S, F, T>(signer: S, func: F) -> Response
 where
     S: Fn(&[u8]) -> String,
     F: FnOnce() -> Result<T, warp::Rejection> + Send + 'static,
     T: Serialize + Send + 'static,
 {
-    warp_utils::task::blocking_task(func)
+    let result = warp_utils::task::blocking_task(func)
         .await
         .map(|func_output| {
-            let mut response = match serde_json::to_vec(&func_output) {
-                Ok(body) => {
-                    let mut res = Response::new(body);
-                    res.headers_mut()
-                        .insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
-                    res
+            let response_body = match serde_json::to_vec(&func_output) {
+                Ok(body) => body,
+                Err(_) => {
+                    return WarpResponse::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from(vec![]))
+                        .expect("can produce simple response from static values")
```

Comment:
By returning this error early we avoid signing it, which is a regression compared to the previous behaviour. Later if we convert an error to a response with `convert_rejection` we also don't sign that, when we should.

Let's refactor this so that we first produce a `Result<Response, Rejection>` from `warp_utils::task::blocking_task`. Then we run `convert_rejection` to get a `Response`, and then finally we sign the body of that response and append the signature to the headers using `headers_mut`. I.e. re-use this block of code from the old implementation, but only after calling `convert_rejection`.

https://github.com/sigp/lighthouse/blob/c7e978b727ec7dbc62951af5e92e6d5413e4c74e/validator_client/src/http_api/mod.rs#L1254-L1259

Reply:
Thanks for the clarification :green_heart:.It makes more sense now., I've just created a PR following it.